polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs

```diff
@@ -1224,18 +1275,74 @@ impl FragmentChain {
 		}
 	}
 
-	/// Mark a candidate as backed. Return `None` if the candidate is not part of the unconnected
-	/// storage.
-	/// This will trigger a recreation of the best backable chain.
-	pub fn candidate_backed(&self, newly_backed_candidate: &CandidateHash) -> Option<Self> {
-		// Get the storage containing both the backable chain and the unconnected candidates.
-		let mut old_storage = self.as_candidate_storage();
+	/// Mark a candidate as backed. This can trigger a recreation of the best backable chain.
+	pub fn candidate_backed(&mut self, newly_backed_candidate: &CandidateHash) {
+		// Already backed.
+		if self.best_chain.candidates.contains(newly_backed_candidate) {
+			return
+		}
+		let Some(parent_head_hash) = self
+			.unconnected
+			.by_candidate_hash
+			.get(newly_backed_candidate)
+			.map(|entry| entry.parent_head_data_hash)
+		else {
+			// Candidate is not in unconnected storage.
+			return
+		};
+
+		// Mark the candidate hash.
+		self.unconnected.mark_backed(newly_backed_candidate);
 
-		if !old_storage.mark_backed(newly_backed_candidate) {
-			return None
+		// Revert to parent_head_hash
+		if !self.revert_to(&parent_head_hash) {
+			// If nothing was reverted, there is nothing we can do for now.
+			return
 		}
 
-		// Repopulate.
-		Some(Self::populate(self.scope.clone(), old_storage))
+		let mut prev_storage = std::mem::take(&mut self.unconnected);
+
+		// Populate the chain.
+		self.populate_chain(&mut prev_storage);
+
+		// Now that we picked the best backable chain, trim the forks generated by candidates
+		// which are not present in the best chain. We can start trimming from this candidate
+		// onwards.
+		self.trim_uneligible_forks(&mut prev_storage, Some(parent_head_hash));
+
+		// Finally, keep any candidates which haven't been trimmed but still have potential.
+		self.populate_unconnected_potential_candidates(prev_storage);
+	}
+
+	// Revert the best backable chain so that the last candidate will be one outputting the given
+	// `parent_head_hash`. If the `parent_head_hash` is exactly the required parent of the base
+	// constraints (builds on the latest included candidate), revert the entire chain.
+	// Return false if we couldn't find the parent head hash.
+	fn revert_to(&mut self, parent_head_hash: &Hash) -> bool {
+		let mut removed_items = None;
+		if &self.scope.base_constraints.required_parent.hash() == parent_head_hash {
+			removed_items = Some(self.best_chain.clear());
+		}
+
+		if removed_items.is_none() && self.best_chain.by_output_head.contains_key(parent_head_hash)
+		{
+			removed_items = Some(self.best_chain.revert_to_parent_hash(parent_head_hash).collect());
+		}
+
+		let Some(removed_items) = removed_items else { return false };
+
+		// Even if it's empty, we need to return true, because we'll be able to add a new candidate
+		// to the chain.
+		for node in removed_items {
+			let _ = self.unconnected.add_candidate_entry(CandidateEntry {
+				candidate_hash: node.candidate_hash,
```

Comment:
Hmm. Maybe we should have a `CandidateEntry::from(FragmentNode)` instance?

Reply:
I wouldn't add that, as there is no obvious way of telling what `CandidateState` (backed or seconded) to use without additional context

Reply:
Why not: I believe the invariant is that everything that is on a chain is that it was backed?

FragmentNode -> CandidateEntry should be good.


I am assuming here that a `FragmentNode` is an element of `BackedChain` ... although I am mostly inferring this from the code, this invariant is actually not documented.

Reply:
done